solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java

```diff
@@ -118,7 +121,22 @@ private ShardResponse take(boolean bailOnError) {
 
     while (pending.size() > 0) {
       try {
-        Future<ShardResponse> future = completionService.take();
+        SolrRequestInfo reqInfo = SolrRequestInfo.getRequestInfo();
+        Future<ShardResponse> future = null;
+        if(reqInfo != null && reqInfo.getTimeAllowed() > -1 ) {
+          long expireAt = reqInfo.getNOW().getTime() + reqInfo.getTimeAllowed();
+          long currTime = System.currentTimeMillis();
+          if (currTime > expireAt) {
+            cancelAll(true);
+            throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, "Operation timed out");
+          }
+          future = completionService.poll(expireAt - currTime, TimeUnit.MILLISECONDS);
+          if (future == null) {
+            throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, "Operation timed out");
+          }
```

Comment:
This appears to be the case when the http shard executor task hangs during my test. 

The exception thrown here does unblock the client (my load generator in this case), but it appears to slowly increase # of threads hanging at `(InputStreamResponseListener.java:318)` (around 12 threads within an hour) . Though this is not significant, should we consider cancelling all pending tasks just like the other case?

Reply:
Just wondering what is the preference for this type of error? Should we throw a server error here (which appears to propagate a 500 back to my client) or should we be a bit more forgiving and consider that as a shard specific failure and honor `bailOnError` check like https://github.com/fullstorydev/lucene-solr/pull/106/files#diff-ce2580113ae9a9d50f308985229130f094868e83d35a8727c5683e6a2f3f44daR142 ?

From our prod log with `TimePatrol` check and my local test, this usually is only one bad shard that triggers this, should we consider to return a partial result still? :)



From my test case, it usually is just one shard that went bad

Reply:
Can you try adding `shards.tolerant=true` in your load generator to see if the client still sees the error? I believe that's what is used in production too.

Reply:
Tried with http://127.0.0.1:8984/solr/fsloadtest/select?q=*%3A*&timeAllowed=20000&shards.tolerant=true and it's still getting 500 on the client 😢

Reply:
Made an attempt to return partial results https://github.com/fullstorydev/lucene-solr/compare/noble/timeallowed...fullstorydev:patson/timeallowed-proposal?expand=1 . In the experiment, there are various issues trying to return partial results if there's hangs for one (or more) shards:
1. Technically there's no exception, but we can create one with `UNKNOWN` response code
2. With the existing flow, a partial result expects a `ShardResponse` with exception set. However, when it times out, there's no `ShareResponse` from the `Future` as `poll` times out without giving back any completed Future
3. CompletionService is not cleaned up even if the pending futures are cancelled, and the existing code logic will go into `take` once more (normally it exits right the way as there are no pending left for this stage), but in our timeout case, the CompletionService actually still have the hanging Future which is now cancelled, and call `Future.get` on it throws `CancellationException`
4. There are various cases that new code can trigger NPE due to race condition

All of the above are addressed in the experimental code branch. 

Unfortunately, there's one **FATAL** issue with my experimental approach is that the current logic computes the expiry time as `NOW` + timeAllowed, which on first stage (only getting doc IDs) if there's one shard hangs, the timeout would be triggered, we will set the exception and return a response (along with all the other successful shard responses), but once it advances to the stage to retrieve actual docs, it will timeout right the way as expiry time is still `NOW` + timeAllowed, which should be in the past already. Therefore the doc retrieval request can never run.

This leads me to think that, it's very tricky to get partial docs result right when there was a timeout. Perhaps we shouldn't even try if this is supposed to be an edge condition?

Reply:
The objective of the PR was to fail the request and return an error to the client if the time out is exceeded. Ideally, we should fix the jetty client code. But, even then it's a much better situation than running out of threads because of this

Reply:
Yes, I agree. Timing out is a good protection for us and throwing exception is simpler than trying to do partial response, which is tricky.

Should we consider calling `cancelAll` here as well?

Reply:
And something kinda minor, but perhaps we don't need https://github.com/fullstorydev/lucene-solr/pull/106/files#diff-ce2580113ae9a9d50f308985229130f094868e83d35a8727c5683e6a2f3f44daR128-R132 . If `currTime > expireAt`, then `completionService.poll(expireAt - currTime, TimeUnit.MILLISECONDS);` should return immediately with `null`, hence the rest of the handling should work as is. This eliminates one branch :)

Reply:
re-ping @noblepaul , should we consider `cancelAll` when `completionService.poll` line 134 - 136 returns null, otherwise the hanging thread will still be running

Reply:
@patsonluk 

I think we should `cancelAll()` if `poll()` returns `null` as  `null` is returned after it timed out